Sources/TensorFlow/Operators/NN.swift

```diff
@@ -90,6 +90,8 @@ public func conv2D<Scalar: TensorFlowFloatingPoint>(
     padding: Padding = .valid,
     dilations: (Int, Int, Int, Int) = (1, 1, 1, 1)
 ) -> Tensor<Scalar> {
+    assert(input.shape.rank == 4, "input must have rank 4")
```

Comment:
Is there a reason to use an `assert` instead of `precondition` ?
Also would it make sense to put the filter precondition in the initialization? Since each precondition adds some overhead, we might be reducing some by checking for filter at init and input at call.
WDYT?

Reply:
Please use `precondition` instead of `assert`.

Reply:
@rxwei perfect. I changed the assertions for preconditions. 
@Shashi456 maybe I am missing something, but as the function receives both of the parameters, you need to check the precondition every time or there is a risk of failing without error.